### PR TITLE
Add support for per-item commit characters in LSP completion

### LINKED ISSUE #44761: [LSP] Define constant commit characters
Certain characters should always commit in an LSP completion session.  They should be set on the CompletionOptions we pass in during Initialize

dot, space, enter, parens, braces, brackets

https://github.com/dotnet/roslyn/blob/master/src/Features/LanguageServer/Protocol/Handler/Initialize/InitializeHandler.cs#L47

### LINKED ISSUE #44893: Verify that completion tests using VerifyProviderCommitAsync are valid
The completion test [TestAllowTypingTypeParameters](http://sourceroslyn.io/#Microsoft.CodeAnalysis.VisualBasic.EditorFeatures.UnitTests/Completion/CompletionProviders/CrefCompletionProviderTests.vb,bc56c6c504760118) tests an impossible scenario (i.e. `IList(Of T)` wouldn't be in the list after typing `IList(O`. It would be good to make sure that the other tests using VerifyProviderCommitAsync are testing valid scenarios.

### PR DESCRIPTION
**Pull request by allisonchou**
- Adds support for per-item commit characters and constant commit characters. (Resolves #44761, https://devdiv.visualstudio.com/DevDiv/_queries/edit/1133648)
- Removes unnecessary `textTypedSoFar` parameter in the method that checks the validity of commit characters, and updates applicable tests (this is probably around half of the PR changes). Additionally, some of these tests previously tested impossible scenarios, so they have been modified. (Resolves #44893)
- Fixed issues with completion list showing up in undesired locations (https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1168461/, https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1178022/)

### CODE REVIEW

File: src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/CommitManager.cs [hunk 1]
```diff
@@ -316,12 +316,6 @@ internal static bool IsCommitCharacter(CompletionRules completionRules, Completi
                 }
             }
 
-            // general rule: if the filtering text exactly matches the start of the item then it must be a filter character
-            if (Helpers.TextTypedSoFarMatchesItem(item, textTypedSoFar))
-            {
-                return false;
-            }
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/CommitManager.cs:323]: Why was this removed?  it seems like something intentional.

Comment (allisonchou) [reply src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/CommitManager.cs:323]: There's two main reasons:
- The check doesn't seem to affect anything. After running all the completion tests, the only test that failed when the check was removed was a test that [tested an impossible scenario](https://github.com/dotnet/roslyn/pull/45099/files#diff-6924c1ebe069be209b3f9f77c73d9e13R354).
- We are hoping to use [similar logic](https://github.com/dotnet/roslyn/pull/45099/files#diff-0735f44f689a499a4e9ecb6aa4a5c2deR120) for LSP commit characters. LSP completion doesn't support text typed so far (and I don't believe there are plans to do so). So I think making the logic as similar as possible would be ideal.

Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/CommitManager.cs:323]: can you validate some typing scenarios that involve items in teh list that have characters that aren't identifier chars?  i.e. making sure that `@` and `:` and `=` and ` ` are handled properly for c#?  (and similar for VB).

`@` appears for identifiers that conflict with keyword names.  i.e. `@int`
`:` appears for named args
` = ` appears for attribute properties.

Comment (allisonchou) [reply src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/CommitManager.cs:323]: It looks like you're right and some of those scenarios currently don't work. :(

- `@`: committing an item that starts with `@` can result in duplicate `@@` - issue filed: https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1142478
- `:` testing `:` is currently impossible due to this bug which crashes LSP: https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1132899/
- `=`: wasn't working properly before, but just pushed a fix.

I appreciate the validation suggestions!

Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/CommitManager.cs:323]: For all the cases, we should have tests btw :)

Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/CommitManager.cs:323]: Note: i would consider these issues current blockers on this as they break muscle memory.

Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/CommitManager.cs:323]: `:` crashes LSP...????????????????????? :)

**Comment by allisonchou** [comment]
Marking as blocked for now until the following bugs are resolved:
https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1142478
https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1132899/

File: src/Features/LanguageServer/Protocol/Handler/Completion/CompletionResolveData.cs [hunk 2]
```diff
@@ -13,5 +14,7 @@ internal class CompletionResolveData
         public Position Position { get; set; }
 
         public string DisplayText { get; set; }
+
+        public CompletionTrigger CompletionTrigger { get; set; }
```
Comment (allisonchou) [inline src/Features/LanguageServer/Protocol/Handler/Completion/CompletionResolveData.cs:18]: Anyone know whether it's OK to pass Roslyn types to LSP? CompletionTrigger is a struct defined in Roslyn.
Alternatively, instead of using CompletionTrigger here, we could instead pass in the trigger character and LSP trigger kind, but this would later require a conversion and the allocation of a CompletionTrigger object in CompletionResolveHandler.

File: src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs [hunk 3]
```diff
@@ -58,44 +72,116 @@ public override async Task<LSP.CompletionItem[]> HandleRequestAsync(LSP.Completi
...
+                CompletionItem item,
+                CompletionTrigger completionTrigger) where TCompletionItem : LSP.CompletionItem, new()
+            {
+                var completeDisplayText = item.DisplayTextPrefix + item.DisplayText + item.DisplayTextSuffix;
+                var completionItem = new TCompletionItem
                 {
-                    Label = item.DisplayTextPrefix + item.DisplayText + item.DisplayTextSuffix,
-                    InsertText = item.Properties.ContainsKey("InsertionText") ? item.Properties["InsertionText"] : item.DisplayText,
+                    Label = completeDisplayText,
+                    InsertText = item.Properties.ContainsKey("InsertionText") ? item.Properties["InsertionText"] : completeDisplayText,
```
Comment (allisonchou) [inline src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs:122]: Changed from `item.DisplayText` to `completeDisplayText` to address the `=` case (discussed in https://github.com/dotnet/roslyn/pull/45099#discussion_r439699963).

File: src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs [hunk 4]
```diff
@@ -58,44 +72,116 @@ public override async Task<LSP.CompletionItem[]> HandleRequestAsync(LSP.Completi
                 .WithChangedOption(CompletionServiceOptions.DisallowAddingImports, true);
 
             var completionService = document.Project.LanguageServices.GetRequiredService<CompletionService>();
-            var list = await completionService.GetCompletionsAsync(document, position, options: completionOptions, cancellationToken: cancellationToken).ConfigureAwait(false);
+
+            // TO-DO: More LSP.CompletionTriggerKind mappings are required to properly map to Roslyn CompletionTriggerKinds.
+            // https://dev.azure.com/devdiv/DevDiv/_workitems/edit/1178726
+            var triggerCharacter = char.Parse(request.Context.TriggerCharacter);
+            var triggerKind = ProtocolConversions.LSPToRoslynCompletionTriggerKind(request.Context.TriggerKind);
+            var completionTrigger = new CompletionTrigger(triggerKind, triggerCharacter);
```
Comment (allisonchou) [inline src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs:80]: We need to pass in `completionTrigger` to `GetCompletionsAsync` to ensure the completion list does not appear in invalid locations (relevant bug is https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1178022/).

File: src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs [hunk 5]
```diff
@@ -40,6 +42,18 @@ public override async Task<LSP.CompletionItem[]> HandleRequestAsync(LSP.Completi
                 return Array.Empty<LSP.CompletionItem>();
             }
 
+            // C# and VB share the same LSP language server, and thus share the same default trigger characters.
+            // The '{' character, while a trigger character in VB, is not a trigger character in C#, so we must
+            // handle it in a special case.
+            if (request.Context.TriggerCharacter == "{")
```
Comment (allisonchou) [inline src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs:48]: Relevant bug: https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1168461/

File: src/EditorFeatures/DiagnosticsTestUtilities/CodeActions/AbstractCodeActionOrUserDiagnosticTest.cs [hunk 6]
```diff
@@ -455,31 +455,6 @@ static IEnumerable<Location> GetUnnecessaryLocations(Diagnostic diagnostic)
             }
         }
 
-        private static void TestDiagnosticTags(
```
Comment (allisonchou) [inline src/EditorFeatures/DiagnosticsTestUtilities/CodeActions/AbstractCodeActionOrUserDiagnosticTest.cs:458]: This doesn't have anything to do with the commit character PR - I somehow just had this file open and noticed this unused method.

File: src/EditorFeatures/CSharpTest/Completion/CompletionProviders/CrefCompletionProviderTests.cs [hunk 7]
```diff
@@ -264,7 +264,7 @@ public async Task CommitUnqualifiedTypes()
         {
             var text = @"
 using System.Collections.Generic;
-/// <see cref=""List{T}.$$""/>
```
Comment (allisonchou) [inline src/EditorFeatures/CSharpTest/Completion/CompletionProviders/CrefCompletionProviderTests.cs:267]: Originally tested an impossible scenario (see https://github.com/dotnet/roslyn/pull/45099/files#r439681997).

File: src/EditorFeatures/CSharpTest/Completion/CompletionProviders/CrefCompletionProviderTests.cs [hunk 8]
```diff
@@ -273,7 +273,7 @@ class C { }
 /// <see cref=""List{T}.Enumerator ""/>
 class C { }
 ";
-            await VerifyProviderCommitAsync(text, "Enumerator", expected, ' ', "Enum");
+            await VerifyProviderCommitAsync(text, "Enumerator", expected, ' ');
```
Comment (allisonchou) [inline src/EditorFeatures/CSharpTest/Completion/CompletionProviders/CrefCompletionProviderTests.cs:276]: All test changes in non-LSP test files are in response to the change here: https://github.com/dotnet/roslyn/pull/45099/files#r439293576

**Comment by allisonchou** [comment]
@CyrusNajmabadi @davidwengier @dibarbet 
I believe this PR is ready for review again as the previous blocking issues seem to have been fixed.
I left some comments throughout the PR, but one aspect I'm still unsure about is testing commit characters (re: Cyrus' comment here - https://github.com/dotnet/roslyn/pull/45099#discussion_r441167743). Since we don't control the actual committing of characters, I don't think there's currently a good way to test. If anyone has any thoughts, please feel free to share.

File: src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs (hunk 4 above)
Comment (allisonchou) [reply src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs:80]: (Planning on hopefully adding a test for this, but it was a bit complicated to get working properly and I wanted to get the rest of the PR out first.)

File: src/Features/LanguageServer/Protocol/Handler/Completion/CompletionResolveData.cs (hunk 2 above)
Comment (davidwengier) [reply src/Features/LanguageServer/Protocol/Handler/Completion/CompletionResolveData.cs:18]: Anything you pass just gets serialized, and would need to be deserialized on the other end, so if the structs are the same shape and the same DataContract attributes it would work... but I'm guessing this doesn't have that :)

File: src/Features/LanguageServer/Protocol/Extensions/ProtocolConversions.cs [hunk 9]
```diff
@@ -65,6 +65,16 @@ internal static class ProtocolConversions
...
         };
 
+        // TO-DO: More LSP.CompletionTriggerKind mappings are required to properly map to Roslyn CompletionTriggerKinds.
+        // https://dev.azure.com/devdiv/DevDiv/_workitems/edit/1178726
+        public static Completion.CompletionTriggerKind LSPToRoslynCompletionTriggerKind(LSP.CompletionTriggerKind triggerKind)
+            => triggerKind switch
+            {
+                LSP.CompletionTriggerKind.Invoked => Completion.CompletionTriggerKind.Invoke,
+                LSP.CompletionTriggerKind.TriggerCharacter => Completion.CompletionTriggerKind.Insertion,
+                _ => throw new ArgumentException($"LSP.CompletionTriggerKind {triggerKind} unexpected.")
```
Comment (dibarbet) [inline src/Features/LanguageServer/Protocol/Extensions/ProtocolConversions.cs:75]: I don't know if we should necessarily throw but instead log something and continue.  It's possible that new trigger kinds get added underneath us (for example, the spec defines a kind not listed here, TriggerForIncompleteCompletions)

File: src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs [hunk 10]
```diff
@@ -40,6 +42,18 @@ public override async Task<LSP.CompletionItem[]> HandleRequestAsync(LSP.Completi
...
+            {
+                var syntaxRoot = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
```
Comment (dibarbet) [inline src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs:50]: This can just be document.Project.Language, no need to get the tree.

File: src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs (hunk 5 above)
Comment (dibarbet) [inline src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs:48]: I know I suggested this initially, but I wonder if instead we could look at the trigger characters per language.  Here we import them and ignore language differences - http://sourceroslyn.io/#Microsoft.CodeAnalysis.LanguageServer.Protocol/Handler/Initialize/InitializeHandler.cs,33 - but we could separate them out instead.  THen in here we can check that the character is applicable for the document language.

File: src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs [hunk 11]
```diff
@@ -58,44 +72,116 @@ public override async Task<LSP.CompletionItem[]> HandleRequestAsync(LSP.Completi
...
-                    var vsCompletionItem = CreateCompletionItem<LSP.VSCompletionItem>(request, item);
+                    var vsCompletionItem = CreateCompletionItem<LSP.VSCompletionItem>(request, item, completionTrigger);
                     vsCompletionItem.Icon = new ImageElement(item.Tags.GetFirstGlyph().GetImageId());
                     return vsCompletionItem;
                 }
                 else
                 {
-                    var roslynCompletionItem = CreateCompletionItem<RoslynCompletionItem>(request, item);
+                    var roslynCompletionItem = CreateCompletionItem<RoslynCompletionItem>(request, item, completionTrigger);
                     roslynCompletionItem.Tags = item.Tags.ToArray();
```
Comment (dibarbet) [inline src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs:108]: I think roslyn completion item can be removed - it was used in the old liveshare implementation.  If vscompletionItem is not supported, we should probably just return a regular CompletionItem

File: src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs [hunk 12]
```diff
@@ -58,44 +72,116 @@ public override async Task<LSP.CompletionItem[]> HandleRequestAsync(LSP.Completi
...
+                            {
+                                if (!commitCharacters.Contains(c))
+                                {
+                                    commitCharacters.Add(c);
+                                }
+                            }
+                            continue;
+
+                        case CharacterSetModificationKind.Remove:
+                            foreach (var c in rule.Characters)
```
Comment (dibarbet) [inline src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs:167]: It seems like order here matters, if the remove comes before the add, then the characters to remove might not exist - or does this only modify the default set?

File: src/Features/LanguageServer/Protocol/Handler/Completion/CompletionResolveData.cs (hunk 2 above)
Comment (dibarbet) [reply src/Features/LanguageServer/Protocol/Handler/Completion/CompletionResolveData.cs:18]: Yeah it needs to be serializable - it looks like it's just an enum and a char and streamjsonrpc might just serialize it.

Note that this does not need to be de-serialized on the client, it's pass through back to the server to de-serialize on resolve.

File: src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs (hunk 12 above)
Comment (allisonchou) [reply src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs:177]: I believe it only modifies the default set - the code here is modeled after the commit character behavior in local Roslyn: https://github.com/dotnet/roslyn/blob/master/src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/CommitManager.cs#L300

File: src/Features/LanguageServer/Protocol/Handler/Completion/CompletionResolveData.cs (hunk 2 above)
Comment (allisonchou) [reply src/Features/LanguageServer/Protocol/Handler/Completion/CompletionResolveData.cs:18]: Hmm I debugged through and it does look like the resolve handler is being deserializing the enum/char fine... so hopefully this approach is OK?

Comment (dibarbet) [reply src/Features/LanguageServer/Protocol/Handler/Completion/CompletionResolveData.cs:18]: The only risk I see is if someone adds a new property that is not easily serializable or de serializable.  Because of that I usually define my own types to avoid that risk, but I'm open to other opinions.

File: src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs [hunk 13]
```diff
@@ -26,10 +28,23 @@ namespace Microsoft.CodeAnalysis.LanguageServer.Handler
     [ExportLspMethod(LSP.Methods.TextDocumentCompletionName)]
     internal class CompletionHandler : AbstractRequestHandler<LSP.CompletionParams, LSP.CompletionItem[]>
     {
+        private readonly string[] _csTriggerCharacters;
+        private readonly string[] _vbTriggerCharacters;
```
Comment (dibarbet) [inline src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs:32]: Let's make these a set so we don't have to worry about duplicates and contains is more straightforward

File: src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs [hunk 14]
```diff
@@ -58,44 +83,125 @@ public override async Task<LSP.CompletionItem[]> HandleRequestAsync(LSP.Completi
                 .WithChangedOption(CompletionServiceOptions.DisallowAddingImports, true);
 
             var completionService = document.Project.LanguageServices.GetRequiredService<CompletionService>();
-            var list = await completionService.GetCompletionsAsync(document, position, options: completionOptions, cancellationToken: cancellationToken).ConfigureAwait(false);
+
+            // TO-DO: More LSP.CompletionTriggerKind mappings are required to properly map to Roslyn CompletionTriggerKinds.
+            // https://dev.azure.com/devdiv/DevDiv/_workitems/edit/1178726
+            var triggerCharacter = char.Parse(request.Context.TriggerCharacter);
```
Comment (dibarbet) [inline src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs:89]: Probably can move this up to parse the trigger char once.

File: src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs [hunk 15]
```diff
@@ -58,44 +83,125 @@ public override async Task<LSP.CompletionItem[]> HandleRequestAsync(LSP.Completi
...
+            {
+                var commitCharacterRules = item.Rules.CommitCharacterRules;
+
+                // If the item doesn't have any special rules, just use the default commit characters.
+                if (commitCharacterRules.IsEmpty)
+                {
+                    return null;
+                }
+
+                using var _ = ArrayBuilder<char>.GetInstance(out var commitCharacters);
```
Comment (dibarbet) [inline src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs:160]: I think this can be switched to a set

File: src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs [hunk 16]
```diff
@@ -58,44 +83,125 @@ public override async Task<LSP.CompletionItem[]> HandleRequestAsync(LSP.Completi
...
+                commitCharacters.AddRange(CompletionRules.Default.DefaultCommitCharacters);
+                foreach (var rule in commitCharacterRules)
+                {
+                    switch (rule.Kind)
+                    {
+                        case CharacterSetModificationKind.Add:
+                            foreach (var c in rule.Characters)
```
Comment (dibarbet) [inline src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs:167]: with a set you can just do set.UnionWith(characters)

File: src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs [hunk 17]
```diff
@@ -58,44 +83,125 @@ public override async Task<LSP.CompletionItem[]> HandleRequestAsync(LSP.Completi
...
+                            {
+                                if (!commitCharacters.Contains(c))
+                                {
+                                    commitCharacters.Add(c);
+                                }
+                            }
+                            continue;
+
+                        case CharacterSetModificationKind.Remove:
+                            foreach (var c in rule.Characters)
```
Comment (dibarbet) [inline src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs:177]: with set this can become exceptwith

File: src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs (hunk 12 above)
Comment (dibarbet) [reply src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs:177]: Interesting - it looks like that exits out on Add if the char matches, on Remove if the char matches, or exits immediately on replace.  We should try and keep that behavior as best as possible.

We can continue on Add and remove - we shouldn't ever get into a situation where we remove something previously added (since it only acts on the default set).  But let's exit early on Replace as that could change the default set, same as how it's done there.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Log/FunctionId.cs [hunk 18]
```diff
@@ -493,5 +493,7 @@ internal enum FunctionId
         DependentTypeFinder_FindAndCacheDerivedClassesAsync = 430,
         DependentTypeFinder_FindAndCacheDerivedInterfacesAsync = 431,
         DependentTypeFinder_FindAndCacheImplementingTypesAsync = 432,
+
+        LSPCompletion_MissingLSPCompletionTriggerKind = 433,
```
Comment (dibarbet) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Log/FunctionId.cs:497]: this will need to be GDPR approved, ping me when this is merged if you don't know how

Comment (allisonchou) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Log/FunctionId.cs:497]: Will do! (I think you've actually shared the link to the GDPR telemetry site with me before, but I've lost it. 😞 )

Also, thanks for the review - switching from using arrays to sets really cleaned up the code in some places.

File: src/Features/LanguageServer/Protocol/Handler/Completion/CompletionResolveData.cs (hunk 2 above)
Comment (allisonchou) [reply src/Features/LanguageServer/Protocol/Handler/Completion/CompletionResolveData.cs:18]: Can we possibly make a note in CompletionTrigger's summary saying that if a new property is added, it should be serializable or LSP needs to be updated?
I think we have something similar for semantic tokens where we include a note saying that if the type/class is updated, LSP should also be updated.

File: src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs [hunk 19]
```diff
@@ -26,10 +29,23 @@ namespace Microsoft.CodeAnalysis.LanguageServer.Handler
...
+        public CompletionHandler(
+            ILspSolutionProvider solutionProvider,
+            [ImportMany] IEnumerable<Lazy<CompletionProvider, CompletionProviderMetadata>> completionProviders)
+            : base(solutionProvider)
         {
+            var csCompletionProviders = completionProviders?.Where(lz => lz.Metadata.Language == LanguageNames.CSharp).ToImmutableArray();
+            var vbCompletionProviders = completionProviders?.Where(lz => lz.Metadata.Language == LanguageNames.VisualBasic).ToImmutableArray();
+
+            _csTriggerCharacters = csCompletionProviders.HasValue ? csCompletionProviders.Value.SelectMany(
+                lz => GetTriggerCharacters(lz.Value)).Distinct().Select(c => c.ToString()).ToImmutableHashSet() : ImmutableHashSet.Create<string>();
```
Comment (dibarbet) [inline src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs:46]: I *think* you don't need to do distinct here, ToImmutable should just add the unique (but check to make sure)

File: src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs [hunk 20]
```diff
@@ -26,10 +29,23 @@ namespace Microsoft.CodeAnalysis.LanguageServer.Handler
...
+
         [ImportingConstructor]
         [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
-        public CompletionHandler(ILspSolutionProvider solutionProvider) : base(solutionProvider)
+        public CompletionHandler(
+            ILspSolutionProvider solutionProvider,
+            [ImportMany] IEnumerable<Lazy<CompletionProvider, CompletionProviderMetadata>> completionProviders)
+            : base(solutionProvider)
         {
+            var csCompletionProviders = completionProviders?.Where(lz => lz.Metadata.Language == LanguageNames.CSharp).ToImmutableArray();
```
Comment (dibarbet) [inline src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs:42]: I don't think you need the ? here

Can probably just do completionProviders.Where(language name equal).SelectMany(GetTriggerChars).Select(tostring).ToImmutableHashSet()

File: src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs [hunk 21]
```diff
@@ -40,6 +56,17 @@ public override async Task<LSP.CompletionItem[]> HandleRequestAsync(LSP.Completi
                 return Array.Empty<LSP.CompletionItem>();
             }
 
+            // C# and VB share the same LSP language server, and thus share the same default trigger characters.
+            // We need to ensure the trigger character is valid in the document's language. For example, the '{'
+            // character, while a trigger character in VB, is not a trigger character in C#.
+            var triggerCharacter = char.Parse(request.Context.TriggerCharacter);
+            if (request.Context.TriggerKind != LSP.CompletionTriggerKind.Invoked && !char.IsLetterOrDigit(triggerCharacter) &&
```
Comment (dibarbet) [inline src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs:63]: let's check specifically that it's a triggercharacter invocation instead of !Invoked.  We wouldn't want to check trigger chars if the LSP adds a deletion trigger kind for example

File: src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs [hunk 22]
```diff
@@ -40,6 +56,17 @@ public override async Task<LSP.CompletionItem[]> HandleRequestAsync(LSP.Completi
...
+                ((document.Project.Language == LanguageNames.CSharp && !_csTriggerCharacters.Contains(request.Context.TriggerCharacter)) ||
```
Comment (dibarbet) [inline src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs:64]: Can we extract the C# and VB checks into a helper that takes in the document and trigger char?  Something like IsValidTriggerCharacterForDocument - should make this if statement easier to read

File: src/Features/LanguageServer/Protocol/Handler/Completion/CompletionResolveData.cs (hunk 2 above)
Comment (dibarbet) [reply src/Features/LanguageServer/Protocol/Handler/Completion/CompletionResolveData.cs:18]: that is OK with me

File: src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs (hunk 20 above)
Comment (allisonchou) [reply src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs]: Would this break TypeScript at all? Do they use this CompletionHandler at all (via TypeScriptHandlerShim)?

Comment (dibarbet) [reply src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs]: there should be completion providers anyway since they're part of the vs mef composition.  But speaking of TS, you may need to update the check to not look for matching trigger chars when in a TS document, otherwise the completion there may break.  Should be able to verify by just livesharing on a web project with a TS file.

Comment (allisonchou) [reply src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs]: I see, thanks! I believe the check specifically only checks C#/VB, but I'll double check TS to make sure nothing is broken.

Comment (allisonchou) [reply src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs]: I tested with a Nodejs .ts file via Liveshare and completion seems to be working fine, so I'll go ahead and merge now.
Thanks for the review!
